### ISSUE TITLE
[sparkline] avoid awkward small, uneven grouping, and confusing column stretching

### ISSUE THREAD

**Issue by JobLeonard** [linnarsson-lab/loom-viewer#122]
**TL;DR:** either plot data directly with near-even column size, or group with _at least_ twenty cells. Column width should be increased to be at least 3 pixels wide to minimize interpretation errors.

Plotting small groups (total cells between [1 - 20]x available pixels)
===

Flame maps highlight an inherent issue that we have to resolve when grouping data or plotting it directly. It applies to the other plotters too, but is invisible. _(nb: the data in the examples below is nonsensical to investigate as a flame map and only illustrates the inherent issue with the plotter)_

When there is so little data to plot that we can show it directly, our sparklines look fine (with a caveat I'll get back to):

![image](https://user-images.githubusercontent.com/259840/31226249-b43a3fa0-a9d5-11e7-8ea3-04bcb0da78c3.png)

Similarly, once we start grouping about 10 cells per column or more, things look pretty good too.

Roughly 10 cells:
![image](https://user-images.githubusercontent.com/259840/31226812-e2ed0b8c-a9d7-11e7-8132-2864fa360797.png)

Roughly 100 cells:

![image](https://user-images.githubusercontent.com/259840/31226930-51dfd4d4-a9d8-11e7-9cc6-4a53bc341a5d.png)

However, there is an "awkward transition"-phase between direct plotting and grouped plotting, which is especially visible by flame maps:

![image](https://user-images.githubusercontent.com/259840/31226338-0ab5694a-a9d6-11e7-82d0-8a16464ccabe.png)

In this example, most groups contain two datapoints, but occassionally we see three. It leads to an ugly, difficult to read flame map. (Another issue specific to flame-maps is that at this scale, the horizontal gradient dominates over the vertical one, making reading non-intuitive)

This issue is not limited to flame maps however: _every one of our plotters_ is grouped like this.

For bar plotters and heat-maps, the difference between averaging two or three values is big, introducing unpredictable biases and artifacts to our plots, and box plots do not have enough statistical power yet to be useful at this stage:

![image](https://user-images.githubusercontent.com/259840/31228574-47df2178-a9de-11e7-9d26-69e2308ba21a.png)

![image](https://user-images.githubusercontent.com/259840/31228624-632e9666-a9de-11e7-95e0-d385a8ae5567.png)

Compare to (roughly) 10 or 100 cells per group:

![image](https://user-images.githubusercontent.com/259840/31226835-f5f4af50-a9d7-11e7-9fde-0451172972b3.png)

![image](https://user-images.githubusercontent.com/259840/31226991-83f83632-a9d8-11e7-9405-d878efb0b7b3.png)

Again: nonsensical data, the point is that min/max, average and quartiles can be told apart.

One way to look at this is that grouping the data is essentially applying statistics. For these statistics to make sense, we need enough data per group, and a small enough variation in group size due to rounding. This rounding error gets smaller as the group increases: groups of 2-3 cells vary 50%, 10-11 cells vary 10%, etc. I just had a brief chat with @gioelelm, to see what would be a good statistical rule of thumb. With the p-value < 0.05 the obvious choice is 20 cells at a minimum per group.

Plotting really small datasets (cells [1/3x - 1]x available pixels);
===

The other end of the spectrum is if we have _less_ data than pixels. that means we can spread the columns over multiple pixels. If we avoid anti-aliasing (and we should, it smears out black-and-white columns into grey mush), that again leads to rounding error. In this case in terms of column _width_.

If our cell count is somewhere between 1/2 and 1x the available pixels, some of the columns will be one pixel wide, and some will be two pixels. ***This is a visual error of 100%***, again without any indication to the viewer when this happens.

**As a rule of thumb, I suggest that columns should be at least 3 pixels wide** (plus retinca correction). that way, the error is reduced to 33%. In pure error terms significant, but the issue here is more one of likelihood of column misinterpretaiton, not of statistical rounding error, so it's not as big of a deal.

_PS: Incidentally this gives an empirical answer to the famous [Sorites paradox](https://en.wikipedia.org/wiki/Sorites_paradox) with "there's a part where it's neither heap nor non-heap but just _awkward, unusable inbetween horribleness_" - I bet that statisticians have reached that conclusion years ago._
